advertisement.js

```diff
@@ -0,0 +1,167 @@
+   * @param {Provider[]|Provider} config.providers
+   * @param {Link} config.entries
+   * @param {Bytes} config.context
+   * @param {Link | null} config.previous
+   * @param {boolean} [config.remove]
+   * @param {boolean} [config.override]
+   */
+  constructor ({ previous, providers, entries, context, remove, override }) {
+    if (!providers || !entries || !context) {
+      throw new Error('providers, entries, and context are required')
+    }
+    if (previous === undefined) {
+      throw new Error('previous must be set. If this is your first advertisement pass null')
+    }
+    this.previous = previous
+    this.providers = Array.isArray(providers) ? providers : [providers]
+    this.entries = entries
+    this.context = context
+    this.remove = remove || false
+    this.override = override || false
+  }
+
+  /**
+   * Convert to IPLD shape, defined by schema.ipldsch
+   * see: https://github.com/ipni/go-libipni/blob/main/ingest/schema/schema.ipldsch
+   */
+  encode () {
+    const provider = this.providers[0]
+    /** @type {import('./schema').AdvertisementOutput} AdvertisementOutput */
+    const value = {
+      Provider: provider.peerId.toCID().toString(),
+      Addresses: provider.addresses.map(a => a.toString()),
+      Signature: new Uint8Array(),
+      Entries: this.entries,
+      ContextID: this.context,
+      Metadata: provider.encodeMetadata(),
+      IsRm: this.remove
+    }
+    if (this.previous) {
+      value.PreviousID = this.previous
+    }
+    if (this.providers.length > 1) {
+      value.ExtendedProvider = {
+        Override: this.override,
+        Providers: this.providers.map(p => ({
+          ID: p.peerId.toCID().toString(),
+          Addresses: p.addresses.map(a => a.toString()),
+          Metadata: p.encodeMetadata(),
+          Signature: new Uint8Array()
+        }))
+      }
+    }
+    return value
+  }
+
+  /**
+   * Convert to IPLD shape and sign
+   */
+  async encodeAndSign () {
+    const ad = this
+    const rootProvider = ad.providers[0]
+    if (ad.remove && ad.providers.length > 1) {
+      // see: https://github.com/ipni/go-libipni/blob/afe2d8ea45b86c2a22f756ee521741c8f99675e5/ingest/schema/envelope.go#L126-L127
+      throw new Error('rm ads are not supported for extended provider signatures')
+    }
+
+    // IPLD shape with empty byte values for Signature properties
+    const value = ad.encode()
+
+    // Advertisement signing
+    const sigBuf = ad.signableBytes()
+    const adSig = await sign(rootProvider.peerId, sigBuf, AD_SIG_CODEC)
+    // root Signature with the signed bytes for ad
+    value.Signature = adSig
+
+    if (value.ExtendedProvider) {
+      const { Providers } = value.ExtendedProvider
+      for (let i = 0; i < Providers.length; i++) {
+        const prov = ad.providers[i]
+        if (prov.peerId.toCID().toString() !== Providers[i].ID) {
+          throw new Error('providers order should match encoded ExtendedProvider.Providers order')
```

Comment:
When would you `encode` without signing? Do we need both? Maybe you could get rid of this check if `encode` always signed...